### ISSUE TITLE
Metalinks cannot be reinstalled if the image was saved while they were uninstalled

### ISSUE THREAD

**Issue by dupriezt** [guillep/pharo-core#8370]
**Describe the bug**
NORMAL: Scenario 1: Install a metalink, uninstall it, reinstall it -> the metalink is still executed
BUG: Scenario 2: Install a metalink, uninstall it, save the image, reinstall it -> the metalink is NOT executed
NORMAL: Scenario 3: Install a metalink, save the image, -> the metalink is still executed

Breakpoints have the same issue.

**To Reproduce**
Here are commented snippets of code to reproduce each scenario. Every snippet starts by compiling a `Greetings` class with two methods: #sayHello and #sayGoodbye. These methods respectively do "self inform: 'hello'" and "self inform: 'goodbye'". In every scenario, a metalink is installed on #sayHello, so that executing this method also executes #sayGoodbye. Every scenario ends by cleaning up: uninstalling the metalink and removing the `Greetings` class from the system.

```Smalltalk
"SCENARIO 1: Install metalink, uninstall metalink, re-install metalink -> metalink is executed"
"Compile base code"
Object subclass: #Greetings instanceVariableNames: '' classVariableNames: ''package: 'MyPackage'.
Greetings compile: 'sayHello self inform: ''Hello'''.
Greetings compile: 'sayGoodbye self inform: ''Goodbye'''.
"Says only hello"
Greetings new sayHello.
"Prepare metalink"
node := (Greetings>>#sayHello) ast body statements first.
metaObject := Greetings new.
metaLink := MetaLink new selector: #sayGoodbye; metaObject: metaObject; yourself.
"Install metalink"
node link: metaLink.
"Says hello and goodblye"
Greetings new sayHello.
"Remove metalink"
metaLink uninstall.
"Says only hello"
Greetings new sayHello.
"Reinstall metalink"
node link: metaLink.
"Says hello and goodbye"
Greetings new sayHello.
"Clean-up: remove metalink and base code"
metaLink uninstall.
(RBRemoveClassRefactoring classNames: { Greetings name}) execute.
```

```Smalltalk
"SCENARIO 2: Install metalink, uninstall metalink, save image while metalink is uninstalled, re-install metalink -> Metalink is NOT executed"
"Compile base code"
Object subclass: #Greetings instanceVariableNames: '' classVariableNames: ''package: 'MyPackage'.
Greetings compile: 'sayHello self inform: ''Hello'''.
Greetings compile: 'sayGoodbye self inform: ''Goodbye'''.
"Says only hello"
Greetings new sayHello.
"Prepare metalink"
node := (Greetings>>#sayHello) ast body statements first.
metaObject := Greetings new.
metaLink := MetaLink new selector: #sayGoodbye; metaObject: metaObject; yourself.
"Install metalink"
node link: metaLink.
"Says hello and goodblye"
Greetings new sayHello.
"Remove metalink"
metaLink uninstall.
"SAVE THE IMAGE"
Smalltalk saveSession.
"Says only hello"
Greetings new sayHello.
"Reinstall metalink"
node link: metaLink.
"SAYS ONLY HELLO"
Greetings new sayHello.
"Clean-up: remove metalink and base code"
metaLink uninstall.
(RBRemoveClassRefactoring classNames: { Greetings name}) execute.
```

```Smalltalk
"SCENARIO 3: Install metalink, save image -> Metalink is executed "
"Compile base code"
Object subclass: #Greetings instanceVariableNames: '' classVariableNames: ''package: 'MyPackage'.
Greetings compile: 'sayHello self inform: ''Hello'''.
Greetings compile: 'sayGoodbye self inform: ''Goodbye'''.
"Says only hello"
Greetings new sayHello.
"Prepare metalink"
node := (Greetings>>#sayHello) ast body statements first.
metaObject := Greetings new.
metaLink := MetaLink new selector: #sayGoodbye; metaObject: metaObject; yourself.
"Install metalink"
node link: metaLink.
"Says hello and goodblye"
Greetings new sayHello.
"SAVE IMAGE"
Smalltalk saveSession.
"Says hello and goodbye"
Greetings new sayHello.
"Clean-up: remove metalink and base code"
metaLink uninstall.
(RBRemoveClassRefactoring classNames: { Greetings name}) execute.
```

**Expected behavior**
A metalink reinstalled after the image was saved while it was not installed should work normally.

**Version information:**
 - OS: Windows 10
- Build information: Pharo-9.0.0+build.1048.sha.cdbcfb1ba14ca80501c21396ca359c0071e83376 (64 Bit)

**Comment by StevenCostiou** [comment]
Could it be a problem with the ast cache @MarcusDenker ?
Because, in the second snippet, if you execute again `node := (Greetings>>#sayHello) ast body statements first.` before reinstalling the link, then it works.

It means the node that was saved in the playground before saving is not the same anymore than the real node of the method.

**Comment by StevenCostiou** [comment]
Actually, it seems that your scenarios all work with the `GTPlayground`, while they do not with `SpPlayground`... could you try out to see if I did not do anything wrong?

That would point to a problem in the new playground variable cache or bindings after saving...

**Comment by StevenCostiou** [comment]
I tried again and it works with GT but fail with Sp.

**Comment by dupriezt** [comment]
There's maybe another issue with playgrounds, but the bug I reported still happens when no playground is involved.
Here's a class file where I reproduced the three scenario as tests.
The class has an instance variable containing a collection, a method to add 1 to the collection, and a method to add 2 to the collection.
The tests execute the "add 2" method, the metalink is installed on that method and calls the "add 1" method beforehand.

The scenarios go like:
- install metalink
- call "add 2"
- assert collection is {1. 2} (True for all three scenarios)
- Scenario1: uninstall metalink, reinstall metalink, call "add 2".-> collection is now {1 . 2. 1. 2.} as it should
- Scenario2: uninstall metalink, save image, reinstall metalink, call "add 2" -> collection is now {1. 2. 2} BUG
- Scenario3: save image, call "add 2" -> collection is now {1. 2. 1. 2.} as it should.
[TestMetaLinkReinstall.zip](https://github.com/pharo-project/pharo/files/5817798/TestMetaLinkReinstall.zip)

The setUp method recompiles the "add 1" and "add 2" methods before every tests, just to be sure no metalink from previous test execution is left behind.

**Comment by StevenCostiou** [comment]
To me the problem seems to be that `node == Greetings>>#sayHello) ast body statements first` returns false, the stored node is not the same as the real method node after saving (why?).

**Comment by MarcusDenker** [comment]
I think the problem is that by un-installing the last link, the ReflectiveMethod is removed. The AST itself is still cached, but when we save we clean the cache.

Then the re-install is done on the node that we still referenced, but that now has nothing to do with the method that is installed.

You get a current node after saving bu re-evaluating:

node := ((self class)>>#addTwoToCollection) ast body statements first.

With that, we now get a node that is from the currently installed method.

**Comment by dupriezt** [comment]
Should a link that get uninstalled keep a reference to the ReflectiveMethod, and reinstall it (somehow) when it is reinstalled?
Is there a way to keep the cache when saving?

I'm not sure how to start to fix this.

**Comment by dupriezt** [comment]
Here's a PR to address this we did with @MarcusDenker . #8448 
The core issue is that MetaLinks store the node they are attached to, and rely on it still being in the ASTCache to work (i.e. be reinstallable). This ASTCache gets cleaned as the image is saved, and the MetaLinks that are uninstalled don't keep their method in the cache because they're uninstalled.
Fixing that would require a lot of thinking.

The case at hand (reinstalling a metaLink) is covered by the MetaLink API: #disable and #enable, but it was buggy.
We fixed it with Marcus, added some tests, and added a comment to MetaLink>>#uninstall saying the user should have a look at #disable and #enable if they want to be sure they can reinstall their metalink
